front/pages/[user]/a/[sId]/execute/index.jsx

```diff
@@ -0,0 +1,403 @@
+                    onChange={(value) => handleValueChange(k, value)}
+                  />
+                </li>
+              ))}
+            </ul>
+            <br /> {/* TODO: no br */}
+            {executionLogs.blockOrder.length ? (
+              <ExecuteOutput
+                executionLogs={executionLogs}
+                expandedByBlockName={outputExpandedByBlockName}
+                onToggleExpand={(blockName) => {
+                  setOutputExpandedByBlockName((prev) => {
+                    const newExpanded = { ...prev };
+                    newExpanded[blockName] = !newExpanded[blockName];
+                    return newExpanded;
+                  });
+                }}
+              />
+            ) : null}
+            <div className="static inset-auto static inset-auto right-0 hidden sm:flex flex-initial items-center pr-2 sm:pr-0">
+              <div className="-mr-2 sm:mr-0">
+                <ActionButton
+                  disabled={isRunning || !isInputDataValid()}
+                  onClick={() => handleRun()}
+                >
+                  <PlayCircleIcon className="-ml-1 mr-1 h-5 w-5 mt-0.5" />
+                  Run
+                </ActionButton>
+              </div>
+            </div>
+          </div>
+        </div>
+      </div>
+    </AppLayout>
+  );
+}
+
+export async function getServerSideProps(context) {
+  const appRes = await fetch(
+    `${URL}/api/apps/${context.query.user}/${context.query.sId}`,
+    {
+      method: "GET",
+      headers: {
+        "Content-Type": "application/json",
+        Cookie: context.req.headers.cookie,
+      },
+    }
+  );
+
+  if (appRes.status === 404) {
+    return {
+      notFound: true,
+    };
+  }
+
+  const app = await appRes.json();
+
+  const savedSpecification = JSON.parse(app.app.savedSpecification || "[]");
+  const inputBlock = savedSpecification.find((block) => block.type === "input");
+  const inputDatasetName = inputBlock.config.dataset;
+
+  const inputDatasetRes = await fetch(
+    `${URL}/api/apps/${context.query.user}/${context.query.sId}/datasets/${inputDatasetName}/latest`,
+    {
+      method: "GET",
+      headers: {
+        "Content-Type": "application/json",
+        Cookie: context.req.headers.cookie,
+      },
+    }
+  );
+  const inputDataset = await inputDatasetRes.json();
+
+  const [specRes] = await Promise.all([
+    fetch(
+      `${URL}/api/apps/${context.query.user}/${context.query.sId}/specification`,
+      {
+        method: "GET",
+        headers: {
+          "Content-Type": "application/json",
+          Cookie: context.req.headers.cookie,
+        },
+      }
+    ),
+  ]);
+
+  if (specRes.status === 404) {
+    return {
+      notFound: true,
+    };
+  }
+
+  const [specData] = await Promise.all([specRes.json()]);
+
+  return {
+    props: {
+      app: app.app,
+      user: context.query.user,
+      ga_tracking_id: GA_TRACKING_ID,
+      specification: specData.specification,
```

Comment:
Ideally we don’t need the specification here but just the hash. It will require a bit more work server side but I think it’s best? wdyt?

Reply:
Actually I wanted to do that but wasn't sure how. Didn't find in the code how to compute the hash. Do you have any pointers ?

Reply:
Here's the change: https://github.com/dust-tt/dust/pull/169/commits/56d0bd82dce030b702561b7fc321d5dfde591431

Apologies if I got this wrong -- my understanding is that the hash is computed on rust side when running the app from the "specification" tab, so it may not be available in the run tab when someone hits `run` (i.e if you go straight to run tab after creating your app without running it).

I believe I have 3 choices:
- do a "saved run" from in the background before actually running against your input -> will be a bit slow and doesn't sound right
- grey out the option of running your app against your input if you haven't done a saved run before (same behaviour as deploy button) -> not sure that the best UX
- use specification hash if available (i.e you did a saved run before), otherwise fallback to passing the whole spec

I went with that last option